basic key derivation for single-account wallet.

dan-da opened the pull request:
This implements basic key derivation using the existing APIs with only a little bit of method signature modification.

This is a useful first step to prove out the concept and demonstrate that tests pass, however it should be emphasized that a more complete implementation would at least support multiple accounts per wallet, each with its own set of addresses.  Monero works that way with a major/minor "path".   Bitcoin has bip32, bip44 etc which define more flexible and advanced path schemes.   Paths are not supported in this PR, so that remains future work/upgrade.    We would ideally do this work before mainnet to avoid upgrade headaches with user's wallets.

This PR also includes dashboard updates including a new Addresses screen to display known addresses in an abbreviated form.   A future commit could display the full address of each when an address row is highilghted and enter key is pressed.

![dashboard-addresses-screen](https://github.com/user-attachments/assets/0ae490e5-d8f7-4194-96b5-28621735868a)

aszepieniec commented on src/models/state/wallet/wallet_state.rs:
```diff
@@ -317,15 +345,18 @@ impl WalletState {
 
                 let spent_utxos = self.scan_for_spent_utxos(&tx.kernel).await;
 
+                let announced_from_addition = self
+                    .scan_addition_records_for_expected_utxos(&tx.kernel.outputs)
+                    .await;
```
This variable name is confusing. This iterator yields `AnnouncedUtxo`s that are a) expected, and b) present in the transaction's list of outputs. In contrast, the next iterator yields `AnnouncedUtxo`s that are a) announced in the transaction's public announcements, and b) present in the transaction's list of outputs.

Consider binding both iterators (not just one) to variables named according to what sets them apart, *i.e.*, point (a).

As a more general commentary (not necessarily suggesting follow-up work): I think the root of the confusion might be the name of the struct `AnnouncedUtxo`. It certainly contains the right information. The name suggests that it can only originate from a public announcement. Expected UTXOs can also be an origin, however.

dan-da replied:
AnnouncedUtxo naming predates this PR and is outside its scope, so will leave it be for now.

I will review the other suggestions, thx.

dan-da replied:
renamed and commented both iterators for clarity.   code is now:

```

                // scan tx for utxo we can claim because we are expecting them (offchain)
                let announced_utxos_from_expected_utxos = self
                    .scan_addition_records_for_expected_utxos(&tx.kernel.outputs)
                    .await;

                // scan tx for utxo with public-announcements we can claim
                let announced_utxos_from_public_announcements =
                    self.scan_for_announced_utxos(&tx.kernel);

                let announced_utxos = announced_utxos_from_public_announcements
                    .chain(announced_utxos_from_expected_utxos)
                    .collect_vec();

```

aszepieniec reviewed (changes requested):
The code looks good to me. I left some remarks inline.

My addresses screen shows duplicates. I think this indicates that something somewhere isn't following expectations.

dan-da commented on src/models/state/wallet/address/address_type.rs:
```diff
@@ -218,23 +227,66 @@ impl ReceivingAddress {
-    /// note: this will fail for Symmetric keys which do not impl bech32m
-    ///       at present.  There is no need to give them out to 3rd parties
-    ///       in a serialized form.
+    /// it would be nice to standardize on a single prefix-len.  6 chars seems a
+    /// bit much.  maybe we could shorten generation prefix to 4 somehow, eg:
+    /// ngkm --> neptune-generation-key-mainnet
```
yes, so the quoted suggestion is simply to standardize on a consistent prefix length, to make things more predictable for end-users.  That does not preclude new address types.

dan-da commented on src/models/state/wallet/address/address_type.rs:
````diff
@@ -218,23 +227,66 @@ impl ReceivingAddress {
+    ///   8 end of address.
+    /// ```
+    ///
+    /// security: note that if this is used on a symmetric key it will display 16 chars
+    /// of the bech32m encoded key.  This seriously reduces the key's strength and it
+    /// may be possible to brute-force it.  In general it is best practice to avoid
+    /// display of any part of a symmetric key.
````
yes, I considered that and perhaps it is the best thing, security wise.

It does introduce an asymmetry between the symmetric display logic and the generation display.   And its not something that should be done only as a hack for dashboard... ie, it should be impl'd for ReceivingAddress.   I will look into a clean way of doing it.  This can have implications for the offchain stuff as well.

aszepieniec replied:
It's not necessary for ReceivingAddress: leaking characters there does expose secret key information. That's by design ;-)

dan-da replied:
> It's not necessary for ReceivingAddress: leaking characters there does expose secret key information. That's by design ;-)

I think you mean "does NOT expose secret key information".  yes?

Also, I think we may be having a miscommunication due to naming.   When I say `ReceivingAddress` I mean exactly `address_type::ReceivingAddress` which abstracts over both `GenerationReceivingAddress` and `SymmetricKey`.  Likewise when I refer to `SpendingKey` I mean `address_type::SpendingKey`, not `GenerationSpendingKey`.

I believe you are thinking of `GenerationReceivingAddress` instead in the above statement.

aszepieniec replied:
> I think you mean "does NOT expose secret key information". yes?

Yes. Oops! I edited it to avoid future confusion.

> I believe you are thinking of GenerationReceivingAddress instead in the above statement.

Yes, correct. I guess I never considered the possibility that symmetric receiving addresses could be semantically loaded. In fact, I'm still skeptical, but for now it suffices to confirm your reading of my message.

dan-da replied:
9f55aad01a232d783b5285ad9ecc10f4d4ee88e8 adds ReceivingAddress::to_display_bech32m_abbreviated() which is now used by the dashboard address screen.  For GenerationReceivingAddress it is the same as before, but for SymmetricKey it internally calls ::privacy_preimage() for a hash before bech32m encoding and abbreviating.   @aszepieniec please confirm if you think this is ok/improvement.

dan-da commented on src/models/state/wallet/wallet_state.rs:
```diff
@@ -225,13 +228,38 @@ impl WalletState {
+                .map(|idx| wallet_secret.nth_generation_spending_key(idx).into())
+                .collect_vec();
+
+        // this is a hack to appease existing tests that bypass
+        // WalletState::next_unused_generation_spending_key() by
+        // calling WalletSecret::nth_generation_spending_key_for_tests().
```
Yeah, this is a bit intricate.  I will unpack... 

So the initial issue I noticed is that many tests call WalletSecret::nth_generation_spending_key_for_tests(0) to directly obtain the 0-th key.  This bypasses the logic in next_unused_generation_spending_key() that increments the wallet's derivation counter and adds the key to known_keys list.

I started a sub-branch to update the tests, but it is not always feasible.    Some of the tests do not involve GlobalState or even WalletState.   Such tests are not really a problem though as they are not testing things that can be affected by not using the counter.

Then I realized that the larger issue is that the application logic uses the 0th generation key for coinbase and pre-mine purposes.   It is assumed to exist (be known).  But if we initially add it to the known-keys list (this hack) and subsequently call next_unused_spending_key() then it is added again, resulting in a dup.  I worked around dups by using a HashSet for known-keys, but that adds some hashing overhead and feels like a bandaid.

I determined that a better/correct fix is to call next_unused_spending_key() inside WalletState::new_from_wallet_secret() immediately after WalletState is created.   

That works correctly with no dups.  The downside is that it results in some tests generating different proofs than we have on the proof server because when the test calls next_unused_spending_key() it obtains the key at index 1 instead of index 0 as it used to.

But it seems objectively correct. 

*Given that coinbase rewards are always paid to key 0 then index 0 should be reserved for coinbase and user-level tx should start at index 1.*  

That's really the important thing to realize here.   and pre-mine rewards are a special case of coinbase rewards.

Then we must ask: why are coinbase rewards always paid to index 0?  Why not derive a addr for each coinbase?    The code presently does it this way because we can avoid acquiring a global-state write-lock (to mutate wallet state) when constructing block template(s) by always using the first key.  There doesn't seem to be a privacy concern because the coinbase ExpectedUtxo is kept offchain by the miner.  There is a comment in mine_loop.rs to this effect.

I should note that any wallet that is not mining and not a pre-mine recipient also has their 0th address skipped for receiving purposes.   This seems fine/correct because any such wallet could begin mining at some point in the future and receive a coinbase reward at key 0.  hence it is *reserved*.

Assuming we go forward with this fix/approach, I may need some assistance getting new proofs onto the proof server(s) since I don't have access.  It's kind of a small detail, but a headache practically for me when testing, pushing to PR, etc.

btw, concretely the discussed approach is to replace the above-referenced "hack" with the following after WalletState is instantiated:

```
        // key 0 is reserved for coinbase and pre-mine.
        // this ensures derivation-index=0 key is known to wallet, so coinbase and premine claiming works.
        // We call it here to avoid acquiring write-lock in mining task when generating tx template.
        // wallets start at index 1 for all other tx.
        if wallet_state.known_generation_keys.is_empty() {
            let _ = wallet_state.next_unused_spending_key(KeyType::Generation).await;
        }
```

aszepieniec replied:
> `   // note: it is Ok to always use the same key here because:`
> `    //  1. if we find a block, the utxo will go to our wallet`
> `    //     and notification occurs offchain, so there is no privacy issue.`
> `    //  2. if we were to derive a new addr for each block then we would`
> `    //     have large gaps since an address only receives funds when`
> `    //     we actually win the mining lottery.`
> `    //  3. also this way we do not have to modify global/wallet state.`
> 
> However @aszepieniec I would like to make sure you agree with the privacy assessment and overall rationale, that it is ok to always mine to the 0 key.

Short answer: yes I agree. :heavy_check_mark: 

Long answer: but it's not obvious that there *isn't* a privacy hazard. In particular, I was afraid that the coinbase UTXO's addition record was going to be identical if they are from the same miner in the same halving epoch. Indeed, the UTXO is identical and the receiver digest is too. But as it turns out, the sender randomness takes the block height into account so we are good.

There is an edge case though, which is when a fork happens to a shorter chain (fewer blocks but more proof-of-work) and the same miner mines block *k*. Well then it will be clear to the outside world that it's the same miner who composed block *k*. I think that is an acceptable compromise.

dan-da replied:
thanks for adding these considerations.    this thread seems to be resolved.

dan-da commented on src/models/state/wallet/mod.rs:
```diff
@@ -201,21 +201,16 @@ impl WalletSecret {
-            "For now we only support one generation address per wallet"
-        );
-
         // We keep n between 0 and 2^16 as this makes it possible to scan all possible addresses
         // in case you don't know with what counter you made the address
```
yep yep

dan-da commented:
> Let's take your classification that an _address_ is something you can spend _to_. Based on the context, I gather that you disagree with my notion that addresses are inherently public. In particular, if change gets routed "back" _to_ your wallet then surely your wallet is in some sense an address.

I would say that traditionally addresses have been both "something you can send to" and "something that is public, ie safe to publish without risk of theft".    However, we have introduced symmetric keys into the mix, which is not "traditional".  And we have abstracted over keytype with `SpendingKey` and `ReceivingAddress`.   In this abstraction, a `ReceivingAddress` is "something you can spend to", but it is not guaranteed to be "something safe to publish".  That's not my opinion or classification... it is simply a description of how the code is working.

> Does the present use of the term "address" satisfy this classification? I would say it does not: you cannot put a symmetric key (or its bech32m encoding) into the "To:" field of the Send interface and expect it to work.

Except that you can.  If you look at the the send() RPC it accepts a `ReceivingAddress` enum which can be either a `SymmetricKey` or `ReceivingAddress` variant.  That is already sufficient to say that RPC client's can send using SymmetricKey.   ie, anyone could write a client that does it.

Going further though: neptune-cli send command calls ReceivingAddress::from_bech32m().   So does the dashboard.  Previously ::from_bech32m() would panic for SymmetricKey, but this PR makes it work... though it was first impl'd in my offchain PR that never got fully merged.

The obvious followup is:  should it work?  This enters into the realm of opinion.   I think you may answer no, and I answer yes.  So I will pro-actively state my opinion/reasons:
1. use cases exist for it.  the most obvious use case is sending payments between wallets owned by same person/entity.  Such payments can be made *onchain* more cheaply/efficiently than equiv onchain payment using GenerationAddress.  And simpler to perform than using an offchain spend/claim, which would be the alternative.
2. It is the cleanest abstraction codepath.   SpendingKey and ReceivingAddress abstract over SymmetricKey and GenerationAddress.  In the future we may add more key types.  It is cleanest if code that interacts with these types treats them all the same instead of special casing for each type.
3. I never like to limit end-users artificially.  We get the functionality basically "for free" due to the abstraction.  So why prohibit its usage?    The reply will be "because symmetric keys are risky and people might accidentally share them", but that is true of all keys.  I would say we should simply document the functionality and risks and let wallet-software makers and end-users make use of it as they choose.
4. We can't really prevent it.  We could disable it in send() rpc and in ::from_bech32m(), yes.   for now. But if someone wants it they can trivially modify the code and release it.
5. it's kind of a neat/novel thing that one can't do in other cryptocurrencies.   why not lean into that?

I feel a little strongly about this, but its not a hill to die on.   So if you don't find my rationale pursuasive then please indicate what changes you would like made in this regard.

>> Only the symmetric items are "Keys", in the sense that each can be spent-from.

> I guess we also disagree about the denotation of the word "key" because from my perspective, keys can be asymmetric or symmetric and in the former case they can be public or secret

I don't disagree with those descriptions of a cryptographic key.   However the cryptocurrency industry has introduced the terminology "address" which is typically used as an alias for public-key and is used for receiving payments.  Even in our code we have terminology "GenerationSpendingKey" and "GenerationReceivingAddress" where the former represent the secret-key and the latter the public-key.  It might have been clearer to name these "GenerationSecretKey" and "GenerationPublicKey".  But regardless the `GenerationSpendingKey` can be used to initiate a tx within the system while the `GenerationReceivingAddress` cannot.

So then we introduced SymmetricKey and abstractions over key-types:
*  `SpendingKey` abstracts over `GenerationSpendingKey` and `SymmetricKey`.
* `ReceivingAddress` abstracts over `GenerationAddress` and `SymmetricKey`.

`SpendingKey` are always secret keys in the sense they can always be spent from.
`ReceivingAddress` are always an address in the sense they can receive funds.

A `SpendingKey` *may* be able to receive funds, but should never be used in that context because it is not guaranteed or intended.
A `ReceivingAddress` *may* be able to spend funds, but should never be used in that context because it is not guaranteed or intended.

Sorry for the verbosity but I am trying to be as clear and factual as I can, to describe where I'm coming from.

> More to the point, the thing you want to display is potentially empty collections of coins that can be spent by the same cryptographic key material. In that sense, the abbreviated receiving address is just the label for the collection and what matters is that the wallet knows the corresponding spending key.

I'm not sure I'm grokking this statement in full.  I don't see that we want to display "potentially empty collections of coins", as I'm not even certain what that means.  Are you using coins to mean Utxos? 

All I'm trying to do is display on the Addresses screen is a list of known keys that can receive funds (pubkey or symmetric), where "known" indicates that the wallet has already derived and "handed-out" that key.   The abbreviated address is simply a shorter form of the bech32m key serialization.    Except that: in latest commits I am testing the abbrev address is a shorter form of a hash of the key, bech32m.

> So in summary:
>
> IMO the information you want to present and display is worthwhile and appropriate.
> IMO the name "address" is inappropriate and confusing for what being displayed.
> IMO the name "spending key" is the essence of what is being displayed, and the encodings are just leakage-free > labels. The word "key" even makes abstraction of symmetric / asymmetric.

I'm not sure what concrete change you are proposing.   Is it to change the name of the screen to "Spending Keys"?  If so, I still don't agree that makes sense, as per above.   But I want to make sure I understand what you would like to see.

aszepieniec commented:
> I'm not sure what concrete change you are proposing.

I was not meaning to propose anything just yet. I just wanted to articulate my position and invite counter-argument. Mission accomplished.

> Except that you can [put a symmetric key (or its bech32m encoding) into the "To:" field of the Send interface and expect it to work]. If you look at the the send() RPC it accepts a ReceivingAddress enum which can be either a SymmetricKey or Generation variant. That is already sufficient to say that RPC client's can send using SymmetricKey. ie, anyone could write a client that does it.
> 
> Going further though: neptune-cli send command calls ReceivingAddress::from_bech32m(). So does the dashboard. Previously ::from_bech32m() would panic for SymmetricKey, but this PR makes it work... though it was first impl'd in my offchain PR that never got fully merged.

Well I did not expect that response O_o. In light of this functionality I would like to suggest a test that runs through this pipeline.

> 1. use cases exist for it. the most obvious use case is sending payments between wallets owned by same person/entity. Such payments can be made onchain more cheaply/efficiently than equiv onchain payment using GenerationAddress. And simpler to perform than using an offchain spend/claim, which would be the alternative.

This is a compelling argument. Yes offchain transactions can do the job also but the reason why you might want the blockchain to store a ciphertext is to robustly back up the UTXO data. And you could achieve that by following up the offchain transaction with an on-chain consolidation transaction from self to self, but that is in fact another use case.

> 5. it's kind of a neat/novel thing that one can't do in other cryptocurrencies. why not lean into that?

It is of course a mis-use hazard. Meaning that unsophisticated users could mistake a not-to-be-made public receiving address for a public one and accidentally reveal or leak it. Potential loss of funds scenario.

Note that the mis-use hazard could be reduced with good naming. In this respect I think "address" scores rather poorly because of its open/public connotation and that's exactly the connotation we want to avoid suggesting if we support this use case.

> I'm not sure I'm grokking this statement in full. I don't see that we want to display "potentially empty collections of coins", as I'm not even certain what that means. Are you using coins to mean Utxos?

Identifying collections with UTXOs is a good first approximation but:
 - some addresses may have not received anything (zero UTXOs)
 - some addresses may have received multiple incoming payments, and these should probably be summed together (multiple UTXOs).

The feature that unifies one collection is the spending key. In principle you can have multiple receiving addresses for the same spending key. In fact, for the lattice-based KEM used in generation addresses, the public key is not uniquely defined with the secret key by default. You have to build in an extra step of derandomization if you want them to be one-to-one, and that's exactly how generation addresses are implemented (*i.e.*, with explicit derandomization).

--

Now that I understand your position better, I find it harder to object to using the word "address", but I'm not convinced it's the best option. I would put forward "key" as a superior alternative.

At this point I think all material arguments have surfaced (if not, all ears). To avoid discussing endlessly about colors and tastes, how about we move forwards as follows?:
 - you give the alternative "key" fair consideration, and then
 - you make the call whether or not to rename anything.

dan-da commented:
> At this point I think all material arguments have surfaced (if not, all ears). To avoid discussing endlessly about colors and tastes, how about we move forwards as follows?:
> 
> * you give the alternative "key" fair consideration, and then
> * you make the call whether or not to rename anything.

thanks, that seems very reasonable.

So I think that, for me, what matters most is consistency in the naming.   I don't feel strongly if its called a key or an address, so long as we are consistent about it both in the code and the user interface.

For example, I could conceive of renaming ReceivingAddress to ReceivingKey, so then we would have SpendingKey and ReceivingKey, which is nicely symmetric, and I think a bit more cryptographically accurate.  And then the "Known Addresses" dashboard screen could be called "Known Receiving Keys" to match.

I don't wish to call the dashboard screen "Keys" when the things populating it are "ReceivingAddress" because it's a big inconsistency -- locally and with usage elsewhere such as neptune-cli. I also don't wish to expand the scope of this PR to include a codebase wide rename.

So I propose/plan that:
1. we leave the Dashboard screen called "Addresses" in the menu and "Known Addresses" in the screen title for this PR.  That accurately reflects what they are called in the code and the rpc calls.
2. Going forward (any time you like) we can rename Addresses throughout the codebase, UI, docs in a dedicated PR.

dan-da commented:
> Please consider adding test cases:
> 
> * All generated addresses and symmetric keys are unique.
> * Restarting the node gives correct counter values.

- [ ] 9f55aad01a232d783b5285ad9ecc10f4d4ee88e8 adds tests:

* known_keys_are_unique
* test derivation_counter_persists_across_restart

dan-da commented:
> Well I did not expect that response O_o. In light of this functionality I would like to suggest a test that runs through this pipeline.

yes good point.  I will add one.

dan-da commented:
> yes good point. I will add one.

- [ ] done. 2bc52441719d2b68d0638f74c5d3bd3056fc9927 refactors the rpc_server::send_to_many() test to operate once for each KeyType variant. Each variant is used as the recipient's keytype.     This means it now tests sending to a symmetric key as well as a generation key.  It will also test sending to any future keytype we might add.  It also performs a bech32m serialize/deserialize roundtrip and verifies it matches the original ReceivingAddress.

dan-da commented:
@aszepieniec I believe I have addressed all comments/concerns and this is ready to merge.   with the caveat that tests are still failing in CI due to not being on proof server yet, but they do pass locally.   I will work with @Sword-Smith to get the proof-server updated.

If you can give this a final lookover and approval, we should be able to get it merged today-ish.

aszepieniec reviewed (approved):
LGTM.
